polytracker/mapping.py

```diff
@@ -4,24 +4,59 @@
 
 from collections import defaultdict
 from pathlib import Path
-from typing import Dict, List, Set, Tuple, Generator
+from typing import Dict, List, Set, Tuple, Iterator
 from tqdm import tqdm
-from mmap import mmap, PROT_READ
 
 from .plugins import Command
-from .taint_dag import TDFile, TDRangeNode, TDSourceNode, TDUnionNode
+from .taint_dag import TDFile, TDNode, TDRangeNode, TDSourceNode, TDUnionNode
 
 
+LabelType = int
 OffsetType = int
+FileOffsetType = Tuple[Path, OffsetType]
 CavityType = Tuple[OffsetType, OffsetType]
 
 
 class InputOutputMapping:
     def __init__(self, f: TDFile):
         self.tdfile: TDFile = f
 
-    def mapping(self) -> Dict[OffsetType, Set[OffsetType]]:
-        raise NotImplementedError()
+    def dfs_walk(
+        self, label: LabelType, seen: Set[LabelType] = set()
+    ) -> Iterator[Tuple[LabelType, TDNode]]:
+        stack = [label]
+        while stack:
+            lbl = stack.pop()
+
+            if lbl in seen:
+                continue
+
+            seen.add(lbl)
+
+            n = self.tdfile.decode_node(lbl)
+
+            yield (lbl, n)
+
+            if isinstance(n, TDSourceNode):
+                continue
+
+            elif isinstance(n, TDUnionNode):
+                stack.append(n.left)
+                stack.append(n.right)
+
+            elif isinstance(n, TDRangeNode):
+                stack.extend(range(n.first, n.last + 1))
+
+    def mapping(self) -> Dict[FileOffsetType, Set[FileOffsetType]]:
+        result: Dict[FileOffsetType, Set[FileOffsetType]] = defaultdict(set)
```

Comment:
```suggestion
        seen = set()
        result: Dict[FileOffsetType, Set[FileOffsetType]] = defaultdict(set)
```

Reply:
I think the `seen` can be reused across the full iteration, that could potentially yield some good speedup. If we've seen a node once we have done all we can with it.